panic while applying a plan

### Terraform Version

```shell
$ terraform version
Terraform v1.8.1-dev
on linux_amd64
+ provider registry.terraform.io/juju/juju v0.13.1

Your version of Terraform is out of date! The latest version
is 1.8.1. You can update by downloading from https://www.terraform.io/downloads.html
```


### Terraform Configuration Files

```terraform
...terraform config...
```


### Debug Output

https://gist.github.com/hmlanigan/d6eb0ce0efffae54681c9d47ae0720a5

### Expected Behavior

terraform should not panic


### Actual Behavior

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: value is marked, so must be unmarked first
goroutine 216 [running]:
runtime/debug.Stack()
	/snap/go/10585/src/runtime/debug/stack.go:24 +0x5e
github.com/hashicorp/terraform/internal/logging.PanicHandler()
	/build/snapcraft-terraform-042b192c7624d48261d394e0860e021d/parts/terraform/build/internal/logging/panic.go:84 +0x18b
panic({0x2e00120?, 0x3d32c90?})
	/snap/go/10585/src/runtime/panic.go:770 +0x132
github.com/zclconf/go-cty/cty.Value.assertUnmarked(...)
	/root/go/pkg/mod/github.com/zclconf/go-cty@v1.14.3/cty/marks.go:141
github.com/zclconf/go-cty/cty.Value.LengthInt({{{0x3d75908?, 0xc000944a10?}}, {0x30d15e0?, 0xc000aca4f8?}})
	/root/go/pkg/mod/github.com/zclconf/go-cty@v1.14.3/cty/value_ops.go:1125 +0x5b
github.com/zclconf/go-cty/cty.UnknownAsNull({{{0x3d75908?, 0xc000944a10?}}, {0x30d15e0?, 0xc000aca4f8?}})
	/root/go/pkg/mod/github.com/zclconf/go-cty@v1.14.3/cty/unknown_as_null.go:41 +0x1b9
github.com/zclconf/go-cty/cty.UnknownAsNull({{{0x3d758d0?, 0xc000944a60?}}, {0x2f38d40?, 0xc000b41710?}})
	/root/go/pkg/mod/github.com/zclconf/go-cty@v1.14.3/cty/unknown_as_null.go:53 +0x3bb
github.com/hashicorp/terraform/internal/terraform.(*NodeAbstractResourceInstance).apply(0xc000b6ad88, {0x3d986a8, 0xc000c393b0}, 0xc0013ba000?, 0xc0013ba000, 0xc0009a5860, {{{{0x0, 0x0}}, {0x0, 0x0}}, ...}, ...)
	/build/snapcraft-terraform-042b192c7624d48261d394e0860e021d/parts/terraform/build/internal/terraform/node_resource_abstract_instance.go:2582 +0x1d55
github.com/hashicorp/terraform/internal/terraform.(*NodeApplyableResourceInstance).managedResourceExecute(0xc000de4800, {0x3d986a8, 0xc000c393b0})
	/build/snapcraft-terraform-042b192c7624d48261d394e0860e021d/parts/terraform/build/internal/terraform/node_resource_apply_instance.go:309 +0xf96
github.com/hashicorp/terraform/internal/terraform.(*NodeApplyableResourceInstance).Execute(0x18?, {0x3d986a8?, 0xc000c393b0?}, 0x70?)
	/build/snapcraft-terraform-042b192c7624d48261d394e0860e021d/parts/terraform/build/internal/terraform/node_resource_apply_instance.go:145 +0x9a
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc0003c7a40, {0x3d986a8, 0xc000c393b0}, {0x78b2c9109200, 0xc000de4800})
	/build/snapcraft-terraform-042b192c7624d48261d394e0860e021d/parts/terraform/build/internal/terraform/graph_walk_context.go:153 +0xbb
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x346ff80, 0xc000de4800})
	/build/snapcraft-terraform-042b192c7624d48261d394e0860e021d/parts/terraform/build/internal/terraform/graph.go:112 +0x62a
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc00120d800, {0x346ff80, 0xc000de4800}, 0xc000e3c7c0)
	/build/snapcraft-terraform-042b192c7624d48261d394e0860e021d/parts/terraform/build/internal/dag/walk.go:384 +0x2d7
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 183
	/build/snapcraft-terraform-042b192c7624d48261d394e0860e021d/parts/terraform/build/internal/dag/walk.go:307 +0xff3

### Steps to Reproduce

`terraform init --upgrade && terraform plan && terraform apply` during development of a new feature for the juju terraform provider plugin.

### Additional Context

I added the following code to a provider while working to get import working:

```
diff --git a/internal/provider/resource_secret.go b/internal/provider/resource_secret.go
index 9a200a5..b9b53f9 100644
--- a/internal/provider/resource_secret.go
+++ b/internal/provider/resource_secret.go
@@ -46,6 +46,8 @@ type secretResourceModel struct {
        SecretId types.String `tfsdk:"secret_id"`
        // Info is the description of the secret. This attribute is optional for all actions.
        Info types.String `tfsdk:"info"`
+       // ID is used during terraform import.
+       ID types.String `tfsdk:"id"`
 }

 // ImportState reads the secret based on the model name and secret name to be
@@ -137,6 +139,11 @@ func (s *secretResource) Schema(_ context.Context, req resource.SchemaRequest, r
                                Description: "The description of the secret.",
                                Optional:    true,
                        },
+                       "id": schema.StringAttribute{
+                               Description: "The ID of the secret.",
+                               Computed:    true,
+                               Optional:    true,
+                       },
                },
        }
 }
```

secretResourceModel is the structure which represents the schema. ID was not used anywhere.

### References

_No response_

--- a/internal/terraform/context_apply2_test.go
+++ b/internal/terraform/context_apply2_test.go
@@ -2917,3 +2917,65 @@ resource "test_object" "obj" {
 	_, diags = ctx.Apply(plan, m, nil)
 	assertNoErrors(t, diags)
 }
+
+func TestContext2Apply_35039(t *testing.T) {
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+resource "test_object" "obj" {
+  list = ["a", "b", "c"]
+}
+`,
+	})
+
+	p := testing_provider.MockProvider{}
+	p.GetProviderSchemaResponse = &providers.GetProviderSchemaResponse{
+		ResourceTypes: map[string]providers.Schema{
+			"test_object": {
+				Block: &configschema.Block{
+					Attributes: map[string]*configschema.Attribute{
+						"output": {
+							Type:     cty.String,
+							Computed: true,
+						},
+						"list": {
+							Type:      cty.List(cty.String),
+							Required:  true,
+							Sensitive: true,
+						},
+					},
+				},
+			},
+		},
+	}
+	p.PlanResourceChangeFn = func(req providers.PlanResourceChangeRequest) providers.PlanResourceChangeResponse {
+		return providers.PlanResourceChangeResponse{
+			PlannedState: cty.ObjectVal(map[string]cty.Value{
+				"output": cty.UnknownVal(cty.String),
+				"list":   req.ProposedNewState.GetAttr("list"),
+			}),
+		}
+	}
+	p.ApplyResourceChangeFn = func(req providers.ApplyResourceChangeRequest) providers.ApplyResourceChangeResponse {
+		return providers.ApplyResourceChangeResponse{
+			// This is a bug, the provider shouldn't return unknown values from
+			// ApplyResourceChange. But, Terraform shouldn't crash in response
+			// to this. It should return a nice error message.
+			NewState: req.PlannedState,
+		}
+	}
+
+	ctx := testContext2(t, &ContextOpts{
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("test"): testProviderFuncFixed(&p),
+		},
+	})
+
+	plan, diags := ctx.Plan(m, states.NewState(), SimplePlanOpts(plans.NormalMode, testInputValuesUnset(m.Module.Variables)))
+	assertNoErrors(t, diags)
+
+	// Just don't crash, should report an error about the provider.
+	_, diags = ctx.Apply(plan, m, nil)
+	if len(diags) != 1 {
+		t.Fatalf("expected exactly one diagnostic, but got %d: %s", len(diags), diags)
+	}
+}

--- a/internal/terraform/node_resource_abstract_instance.go
+++ b/internal/terraform/node_resource_abstract_instance.go
@@ -2438,15 +2438,6 @@ func (n *NodeAbstractResourceInstance) apply(
 	// incomplete.
 	newVal := resp.NewState
 
-	// If we have paths to mark, mark those on this new value we need to
-	// re-check the value against the schema, because nested computed values
-	// won't be included in afterPaths, which are only what was read from the
-	// After plan value.
-	newVal = newVal.MarkWithPaths(afterPaths)
-	if sensitivePaths := schema.SensitivePaths(newVal, nil); len(sensitivePaths) != 0 {
-		newVal = marks.MarkPaths(newVal, marks.Sensitive, sensitivePaths)
-	}
-
 	if newVal == cty.NilVal {
 		// Providers are supposed to return a partial new value even when errors
 		// occur, but sometimes they don't and so in that case we'll patch that up
@@ -2530,6 +2521,15 @@ func (n *NodeAbstractResourceInstance) apply(
 		// bug, we accept this because storing a result here is always a
 		// best-effort sort of thing.
 		newVal = cty.UnknownAsNull(newVal)
+	}
+
+	// If we have paths to mark, mark those on this new value we need to
+	// re-check the value against the schema, because nested computed values
+	// won't be included in afterPaths, which are only what was read from the
+	// After plan value.
+	newVal = newVal.MarkWithPaths(afterPaths)
+	if sensitivePaths := schema.SensitivePaths(newVal, nil); len(sensitivePaths) != 0 {
+		newVal = marks.MarkPaths(newVal, marks.Sensitive, sensitivePaths)
 	}
 
 	if change.Action != plans.Delete && !diags.HasErrors() {

